## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

* Analyze and glean insights from a real dataset using pandas
* Apply pandas for exploratory analysis, information gathering, and discovery
* Demonstrate cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [2]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_format='%d-%b-%y')

# Note - for now, it is okay to ignore the warning about mixed types. 

C:\Users\daehyeon\AppData\Local\Temp\ipykernel_28428\385358853.py:14: DtypeWarning: Columns (0: contbr_zip, 1: receipt_desc, 2: memo_cd) have mixed types. Specify dtype option on import or set low_memory=False.
  contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_format='%d-%b-%y')


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [3]:
# 1a YOUR CODE HERE
print(contrib.shape)

(1125659, 18)


- **1a answer here: The data has **1,125,659 rows and 18 columns**, as expected.** 

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [4]:
# 1b YOUR CODE HERE
print(contrib.columns.tolist())

['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city', 'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation', 'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd', 'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp']


- **1b answer here: All **18 documented columns** are included.** 

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [5]:
# 1c YOUR CODE HERE
print(contrib.head())

     cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26   

- **1c Answer:** The three columns look correct. The candidate IDs and names line up, and the state is `CA`.


**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [7]:
# 1d YOUR CODE HERE
print(contrib['election_tp'].value_counts(dropna=False))


election_tp
P2016    810481
G2016    313746
NaN        1425
P2020         7
Name: count, dtype: int64



- **1d Answer:** `election_tp` shows which election a contribution was meant for. `P` means primary, `G` means general, and the number is the year. The data includes `P2016`, `G2016`, `P2020`, and missing values, and the code format matches the documentation.


**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [9]:
# 1e YOUR CODE HERE
print(contrib.dtypes)


cmte_id                         str
cand_id                         str
cand_nm                         str
contbr_nm                       str
contbr_city                     str
contbr_st                       str
contbr_zip                   object
contbr_employer                 str
contbr_occupation               str
contb_receipt_amt           float64
contb_receipt_dt     datetime64[us]
receipt_desc                    str
memo_cd                         str
memo_text                       str
form_tp                         str
file_num                      int64
tran_id                         str
election_tp                     str
dtype: object


- **1e Answer:** `contbr_zip` is `str`, `contb_receipt_amt` is `float64`, and `contb_receipt_dt` is `datetime64[us]`.


**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [10]:
# 1f YOUR CODE HERE
null_summary = pd.DataFrame({
    'non_null': contrib.notna().sum(),
    'null': contrib.isna().sum()
}).sort_values('non_null', ascending=False)
print(null_summary)

                   non_null     null
cmte_id             1125659        0
cand_id             1125659        0
tran_id             1125659        0
file_num            1125659        0
form_tp             1125659        0
contb_receipt_dt    1125659        0
contb_receipt_amt   1125659        0
contbr_st           1125659        0
contbr_nm           1125659        0
cand_nm             1125659        0
contbr_city         1125633       26
contbr_zip          1125564       95
election_tp         1124234     1425
contbr_occupation   1115260    10399
contbr_employer      967757   157902
memo_text            501148   624511
memo_cd              144268   981391
receipt_desc          15045  1110614


- **1f Answer:** Ten columns have no missing values. `receipt_desc`, `memo_cd`, and `memo_text` are mostly empty and are not needed here, so I would drop them.


**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [11]:
# 1g YOUR CODE HERE
print('Max IDs per name:', contrib.groupby('cand_nm')['cand_id'].nunique().max())
print('Max names per ID:', contrib.groupby('cand_id')['cand_nm'].nunique().max())
print(contrib.loc[contrib['cand_nm'] == 'Sanders, Bernard', 'cand_id'].value_counts())

Max IDs per name: 1
Max names per ID: 1
cand_id
P60007168    407172
Name: count, dtype: int64


- **1g Answer:** Candidate names and IDs match one to one. All 407,172 Sanders records use `P60007168`.


**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [12]:
# 1h YOUR CODE HERE
print(contrib['contbr_st'].value_counts(dropna=False))


contbr_st
CA    1125659
Name: count, dtype: int64


- **1h Answer:** Every value in `contbr_st` is `CA`. There are no rows from outside California.


**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [13]:
# 1i YOUR CODE HERE
duplicate_tran = contrib[contrib['tran_id'].duplicated(keep=False)]
print('Duplicate tran_id after first:', contrib['tran_id'].duplicated().sum())
print('Duplicate composite key:', contrib.duplicated(['cmte_id', 'file_num', 'tran_id']).sum())
print(duplicate_tran[['tran_id', 'cmte_id', 'file_num']].head())

Duplicate tran_id after first: 3454
Duplicate composite key: 0
           tran_id    cmte_id  file_num
57        C3913687  C00575795   1091718
90        C3925472  C00575795   1091718
91        C3937968  C00575795   1091718
174  SA17A.1645838  C00574624   1077664
175  SA17A.1627230  C00574624   1077664


- **1i Answer:** There are 3,454 repeated `tran_id` values after the first occurrence. The `(cmte_id, file_num, tran_id)` combination is unique, so the same `tran_id` can appear in different reports.


**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [14]:
# 1j YOUR CODE HERE
negative = contrib[contrib['contb_receipt_amt'] < 0]
print('Negative rows:', len(negative))
print(negative[['contb_receipt_amt', 'receipt_desc', 'memo_text']].head())

Negative rows: 11896
     contb_receipt_amt              receipt_desc                 memo_text
19              -25.00  REDESIGNATION TO GENERAL  REDESIGNATION TO GENERAL
23             -150.00  REDESIGNATION TO GENERAL  REDESIGNATION TO GENERAL
81              -60.00  REDESIGNATION TO GENERAL  REDESIGNATION TO GENERAL
190            -100.00  REDESIGNATION TO GENERAL  REDESIGNATION TO GENERAL
213             -25.00  REDESIGNATION TO GENERAL  REDESIGNATION TO GENERAL


- **1j Answer:** There are 11,896 negative amounts. The sample rows show refunds and election redesignations, which reverse or adjust earlier contributions.

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [ ]:
# 1k YOUR CODE HERE
start = pd.Timestamp('2014-01-01')
end = pd.Timestamp('2016-06-07')
outside_period = ~contrib['contb_receipt_dt'].between(start, end)
print('Date range:', contrib['contb_receipt_dt'].min(), 'to', contrib['contb_receipt_dt'].max())
print('Outside period:', outside_period.sum())

Date range: 2013-11-05 00:00:00 to 2016-10-19 00:00:00
Outside period: 451372



- **1k Answer:** All dates parsed correctly. They range from November 5, 2013, to October 19, 2016, and 451,372 rows fall outside the stated period.

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here: Yes. The data has 1,125,659 rows and 18 columns.**

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here**: Yes. The data includes the candidate, election type, ZIP code, contribution amount, and date.

**1l.3** What columns are important?

- **1l.3 answer here:I need `cand_nm`, `election_tp`, `contbr_zip`, `contb_receipt_amt`, and `contb_receipt_dt`.** 

**1l.4** What columns can be dropped?

- **1l.4 answer here:**  After filtering, I can keep `zipcode`, `contb_receipt_amt`, and `contb_receipt_dt` and drop the rest.


**1l.5** What are the data problems?

- **1l.5 answer here:** The main issues are missing values, repeated `tran_id` values, negative amounts, invalid ZIP codes, and dates outside the analysis period.


**1l.6** List any assumptions so far:

- **1l.6 answer here:** I include both date endpoints, use the first five ZIP digits, and count each row as one contribution.


***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data. Do each of the following problems sequentially so that at the end the dataframe is filtered and cleaned for the analysis. (that is, use the dataframe answer for 2a to start 2b etc)

**2a.** From the dataset filter out (remove) any election_tp not in the primary election for 2016. Also filter for the primary dates (defined as 1 Jan 2014 to 7 June 2016). Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [16]:
# 2a YOUR CODE HERE
contrib_clean = contrib[
    (contrib['election_tp'] == 'P2016')
    & contrib['contb_receipt_dt'].between('2014-01-01', '2016-06-07')
].copy()
print(contrib_clean.shape)

(668864, 18)


**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [17]:
# 2b YOUR CODE HERE
contrib_clean = contrib_clean[contrib_clean['cand_nm'] == 'Sanders, Bernard'].copy()
print(contrib_clean.shape)

(379284, 18)


**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [18]:
# 2c YOUR CODE HERE
contrib_clean['zipcode'] = contrib_clean['contbr_zip'].astype('string').str.extract(r'^(\d{5})', expand=False)
zip_number = pd.to_numeric(contrib_clean['zipcode'], errors='coerce')
contrib_clean = contrib_clean[zip_number.between(90001, 96162)].copy()
print(contrib_clean.shape)

(379252, 19)


**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [19]:
# 2d YOUR CODE HERE
contrib_clean = contrib_clean[contrib_clean['contb_receipt_amt'] >= 0].copy()
print(contrib_clean.shape)

(376960, 19)


**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [20]:
# 2e YOUR CODE HERE
contrib_clean = contrib_clean[['zipcode', 'contb_receipt_amt', 'contb_receipt_dt']].copy()
print(contrib_clean.shape)

(376960, 3)


- **2e answer here:** I kept only `zipcode`, `contb_receipt_amt`, and `contb_receipt_dt`. The final data has 376,960 rows and 3 columns.


**2f.** List any assumptions that you made up to this point or if there is any filtering you think is needed and why:

NOTE: You can look to see if there are any duplicate rows also - please write an assumption on what you would do with these!

- **2f answer here:** I treated 90001 through 96162 as the California ZIP range and removed only negative amounts. I did not remove any other rows as duplicates because there were no exact duplicate rows.


***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss! That is use the dataframe from 2e above that has done all of the cleaning & filtering steps!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [23]:
# 3a YOUR CODE HERE
zip_summary = contrib_clean.groupby('zipcode').agg(
    count=('contb_receipt_amt', 'size'),
    amount=('contb_receipt_amt', 'sum')
)
print(zip_summary.nlargest(1, 'count'))
print(zip_summary.nlargest(1, 'amount'))

         count     amount
zipcode                  
94110     3799 284,398.05
         count     amount
zipcode                  
94110     3799 284,398.05


- **3a answer here:**   ZIP 94110 had the most contributions at 3,799 and the largest total at $284,398.05.

**3b.** What day(s) of the month do most people donate? (10 points)

In [22]:
# 3b YOUR CODE HERE
day_counts = contrib_clean.groupby(contrib_clean['contb_receipt_dt'].dt.day).size()
print(day_counts.sort_values(ascending=False).head())

contb_receipt_dt
29    21837
31    19485
30    19485
14    16810
9     15133
dtype: int64


- **3b answer here:**  Based on the number of contribution records, the 29th had the highest count at 21,837.

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k